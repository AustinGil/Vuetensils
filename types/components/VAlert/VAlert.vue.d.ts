/**
 * NOTES Concerning :visible prop.
 * Per W3C specifications:
 * https://www.w3.org/TR/wai-aria-practices/#alert
 * --
 * "It is also important to avoid designing alerts that disappear automatically.
 * An alert that disappears too quickly can lead to failure to meet WCAG 2.0 success criterion 2.2.3.
 * Another critical design consideration is the frequency of interruption caused by alerts.
 * Frequent interruptions inhibit usability for people with visual and cognitive disabilities,
 * which makes meeting the requirements of WCAG 2.0 success criterion 2.2.4 more difficult."
 */
declare const _default: {
    name: string;
    model: {
        prop: string;
        event: string;
    };
    props: {
        /**
         * HTML tag used to wrap the component.
         */
        tag: {
            type: StringConstructor;
            default: string;
        };
        /**
         * Determines whether the alert is visible. Also binds with `v-model`.
         */
        modelValue: {
            type: (BooleanConstructor | NumberConstructor)[];
            default: boolean;
        };
        /**
         * Allows a user to dismiss this alert.
         */
        dismissible: BooleanConstructor;
        /**
         * Aria-label that is not visibly, but screen readers will read for the dismiss button.
         */
        dismissLabel: {
            type: (StringConstructor | BooleanConstructor)[];
            default: string;
        };
        /**
         * The transition name if you want to add one.
         */
        transition: {
            type: StringConstructor;
            default: any;
        };
        classes: {
            type: ObjectConstructor;
            default: () => {};
        };
    };
    data: () => {
        dismissed: boolean;
        timerId: any;
    };
    watch: {
        modelValue: {
            handler(visible: any): void;
            immediate: boolean;
        };
    };
    beforeUnmount(): void;
    /** @deprecated */
    beforeDestroy(): void;
    methods: {
        dismiss(): void;
        countdown(): void;
        clearTimer(): void;
    };
};
/**
 * A simple component for notifiying users of specific information. Good for informative snippets, error messages, and more. It can be shown or hidden dynamically, and even supports auto-hiding after a given time.
 */
export default _default;
