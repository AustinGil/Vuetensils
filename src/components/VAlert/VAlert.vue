<template>
  <transition :name="transition">
    <component
      v-if="!dismissed && !!visible"
      :is="tag"
      role="alert"
      class="vts-alert"
    >
      <!-- @slot The default slot content that is shown to the user -->
      <slot />

      <button
        v-if="dismissible"
        @click="dismiss"
        :aria-label="dismissLabel"
        class="vts-alert__dismiss"
      >
        <!-- @slot The dismiss button content -->
        <slot name="dismiss">&times;</slot>
      </button>
    </component>
  </transition>
</template>

<script>
const NAME = "vts-alert"

/**
 * A simple component for notifiying users of specific information. Good for informative snippets, error messages, and more. It can be shown or hidden dynamically, and even supports auto-hiding after a given time.
 */
export default {
  model: {
    prop: "visible",
    event: "update",
  },

  props: {
    /**
     * HTML tag used to wrap the component.
     */
    tag: {
      type: String,
      default: "div",
    },
    /**
     * Determines whether the alert is visible. Also binds with `v-model`.
     */
    visible: {
      type: [Boolean, Number],
      default: true,
    },
    /**
     * Allows a user to dismiss this alert.
     */
    dismissible: {
      type: Boolean,
      default: false,
    },
    /**
     * Aria-label that is not visibly, but screen readers will read for the dismiss button.
     */
    dismissLabel: {
      type: [String, Boolean],
      default: "Dismiss this alert",
    },
    /**
     * The transition name if you want to add one.
     */
    transition: String,
  },

  data: () => ({
    dismissed: false,
    timerId: null,
  }),

  watch: {
    visible: {
      handler(next) {
        if (!!next) {
          this.dismissed = false
        }
        if (typeof this.visible === "number") {
          this.clearTimer() // Clear timers in case this.visible watcher adds multiples
          this.countdown()
        }
      },
      immediate: true,
    },
  },

  methods: {
    dismiss() {
      /**
       * Fired when a user manually dismissed an alert
       * @event dismiss
       * @type { undefined }
       */
      this.$emit("dismiss")
      this.dismissed = true
      if (typeof this.visible === "number") {
        this.$emit("update", 0)
        this.clearTimer()
      } else {
        this.$emit("update", false)
      }
    },

    countdown() {
      if (this.visible <= 0) return

      this.timerId = setTimeout(() => {
        /**
         * Fired whenever the visibility changes. Either through user interaction, or a countdown timer
         * @event update
         * @type { boolean/number }
         */
        this.$emit("update", this.visible - 1)
      }, 1000)
    },

    clearTimer() {
      if (this.timerId) {
        clearInterval(this.timerId)
        this.timerId = null
      }
    },
  },

  beforeDestroy() {
    this.clearTimer()
  },
}
</script>
