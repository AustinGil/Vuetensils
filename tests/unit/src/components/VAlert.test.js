import { mount } from '@vue/test-utils';
import VAlert from '@components/VAlert/VAlert.vue';

describe('VAlert', () => {
  describe('Created', () => {

    it('renders correctly without props', () => {
      const wrapper = mount(VAlert);
      expect(wrapper).toMatchSnapshot();
    });

  });
});
