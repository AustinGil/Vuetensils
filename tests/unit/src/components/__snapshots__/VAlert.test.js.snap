// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VAlert Created Renders correctly 1`] = `
<transition-stub>
  <div role="alert" class="vts-alert">
    <!---->
  </div>
</transition-stub>
`;

exports[`VAlert Created Renders correctly without props 1`] = `
<transition-stub>
  <div role="alert" class="vts-alert">
    <!---->
  </div>
</transition-stub>
`;

exports[`VAlert Created Renders without props 1`] = `
<transition-stub>
  <div role="alert" class="vts-alert">
    <!---->
  </div>
</transition-stub>
`;

exports[`VAlert Created renders correctly without props 1`] = `
<transition-stub>
  <div role="alert" class="vts-alert">
    <!---->
  </div>
</transition-stub>
`;
